redash/query_runner/json_ds.py

```diff
@@ -164,32 +167,80 @@ def run_query(self, query, user):
 
         fields = query.get("fields")
         path = query.get("path")
+        pagination = query.get("pagination", {"type": None})
 
         if isinstance(request_options.get("auth", None), list):
             request_options["auth"] = tuple(request_options["auth"])
         elif self.configuration.get("username") or self.configuration.get("password"):
-            request_options["auth"] = (
-                self.configuration.get("username"),
-                self.configuration.get("password"),
-            )
+            request_options["auth"] = (self.configuration.get("username"), self.configuration.get("password"))
 
         if method not in ("get", "post"):
             raise QueryParseError("Only GET or POST methods are allowed.")
 
         if fields and not isinstance(fields, list):
             raise QueryParseError("'fields' needs to be a list.")
 
-        response, error = self.get_response(query["url"], http_method=method, **request_options)
+        if not isinstance(pagination, dict) or not isinstance(pagination.get("type"), (str, type(None))):
+            raise QueryParseError("'pagination' should be an object with a `type` property")
 
+        results, error = self._get_all_results(query["url"], method, path, pagination, **request_options)
         if error is not None:
             return None, error
 
-        data = json_dumps(parse_json(response.json(), path, fields))
-
+        data = json_dumps(parse_json(results, fields))
         if data:
             return data, None
-        else:
-            return None, "Got empty response from '{}'.".format(query["url"])
+        return None, "Got empty response from '{}'.".format(query["url"])
+
+    def _get_all_results(self, url, method, result_path, pagination, **request_options):
+        """Get all results from a paginated endpoint."""
+        base_url = self.configuration.get("base_url")
+        url = urljoin(base_url, url)
+
+        results = []
+        has_more = True
+        while has_more:
+            response, error = self._get_json_response(url, method, **request_options)
+            has_more = False
+
+            result = _normalize_json(response, result_path)
+            if result:
+                results.extend(result)
+
+                if pagination["type"] == "url":
```

Comment:
I've never seen either of these two pagination patterns. I don't think there's a "standard" either, is there? I imagine every API has its own set of pagination rules that it follows. If that's the case, I'm not sure it makes sense to write in pagination rules, and rather we should emphasize to the end users that querying an API for JSONs directly like this is meant for testing and not large-scale analysis.

Especially since if too much data is returned, it's possible that the worker goes OOM. 😅

Reply:
Some variant of these 2 seem to be the most common, because they don't require specifics of implementation, like having to add to the page number or offset yourself. (The `type` could be named something different if you have a better idea though.) That's why Google outlines the next_page_token/page_token as a best practice. It's just that sometimes the names are different, like Chargebee uses an offset parameter, but rather than having to calculate it, it returns the next offset in the response. Google Analytics, AWS, and K8s also follow that pattern.

Shopify, SalesForce, Facebook's Graph API, and Trino all pass a full URL for the next page in some field of the response, which would be covered by the `url` type. HATEOAS is a standard and does define that the next page would be a link somewhere in the response. Although, at some point, it might be good to use a library instead of `_apply_path_search` to handle more complex cases.

Other types could be added later for `pageNumber` or something, but I don't really have a need for them right now, and that is where it could be pretty difficult to cover different APIs.

I was wondering about the memory thing, but it doesn't look like there's any other way that large result sets are handled elsewhere. The Trino data source does a `fetchall` and then processes all of the results at once. This at least does the `_normalize_json` per page, which could trim things down.

Reply:
Fair points; thanks for elaborating.